scripts/opentitan/autoreg.py

```diff
@@ -0,0 +1,933 @@
+        wo_acc = QEMUAccess('wo')
+        for fpos, item in enumerate(hjson.get('alert_list', [])):
+            fieldargs = [item['name'].upper(), item['desc'], fpos, 1, wo_acc, 0]
+            fields.append(QEMUField(*fieldargs))
+        if not fields:
+            return []
+        regargs = ['ALERT_TEST', 'Alert test', address, wo_acc, fields, 0]
+        return [QEMURegister(*regargs)]
+
+    def _parse_register(self, address: int, reg: dict[str, Any]) \
+            -> Optional[QEMURegister]:
+        name = reg.get('name')
+        if not name:
+            return None
+        if self._ignore_prefix and name.startswith(self._ignore_prefix):
+            return None
+        desc = reg.get('desc')
+        self._log.info('Register: %s', name)
+        fields = self._parse_reg_fields(reg)
+        fields.sort(key=lambda f: f.offset)
+        reset = self._build_reg_reset(fields)
+        access = self._build_reg_access(fields)
+        rname = camel_to_snake_case(name).upper()
+        regargs = [rname, desc, address, access, fields, reset]
+        return QEMURegister(*regargs)
+
+    def _parse_registers(self, address: int, reg: dict[str, Any]) \
+            -> list[QEMURegister]:
+        count = safe_eval(reg['count'], self._parameters)
+        freg = self._parse_register(address, reg)
+        regs = []
+        self._log.info('Generating multireg %s[%d]', freg.name, count)
+        for rpos in range(count):
+            fields = freg.fields if rpos == 0 else []
+            regs.append(
+                freg._replace(name=f'{freg.name}_{rpos}', address=address,
+                              shared_fields=(rpos == 0), fields=fields))
+            address += 4
+        return regs
+
+    def _parse_reg_fields(self, reg: dict[str, Any]) -> list[QEMUField]:
+        fields: list[QEMUField] = []
+        bitfield = 0
+        reg_swaccess = reg.get('swaccess')
+        rname = reg['name']
+        for field in reg.get('fields', []):
+            name = field.get('name')
+            if not name:
+                name = rname
+                if fields:
+                    raise ValueError(f'Multiple anonymous bitfields '
+                                     f'for reg {name}')
+            if self._ignore_prefix and name.startswith(self._ignore_prefix):
+                continue
+            desc = field.get('desc')
+            bits = field.get('bits')
+            if ':' in bits:
+                hibit, lobit = (self._parse_bits(x) for x in bits.split(':'))
+                offset = lobit
+                width = hibit-lobit+1
+            else:
+                offset = self._parse_bits(bits)
+                width = 1
+            bitmask = ((1 << width) - 1) << offset
+            if bitmask & bitfield:
+                raise ValueError(f'Bitfield {name} overlap: '
+                                 f'0x{bitfield:0{self._regwidth}x}, mask '
+                                 f'0x{bitmask:0{self._regwidth}x}')
```

Comment:
Maybe I missed something, but should this not be e.g.

```suggestion
                reg_nibbles = (self._regwidth + 3) / 4
                raise ValueError(f'Bitfield {name} overlap: '
                                 f'0x{bitfield:0{reg_nibbles}x}, mask '
                                 f'0x{bitmask:0{reg_nibbles}x}')
```
Because the `{:0Nx}` python format takes `N` as the number of hex characters i.e. nibbles, whereas the HJSON `regwidth` is a number of bits?

Reply:
You are right; regwidth being a 2**N number, it is safe to divide by 4.